File: utils/Factorization/Factorizator/factorizator.py
import random
import math
import time
from collections import Counter

# ==========================================================
# 大整数因子分解
# 综合算法:
#   1. Trial Division
#   2. Miller-Rabin primality test
#   3. Pollard p-1
#   4. Pollard Rho
# 递归调用,完整分解所有因子
# ==========================================================


# gcd
def gcd(a: int, b: int) -> int:
    while b:
        a, b = b, a % b
    return abs(a)

# Miller-Rabin
def is_prime(n: int, rounds=10) -> bool:

    if n < 2:
        return False

    small_primes = [
        2, 3, 5, 7, 11, 13, 17, 19,
        23, 29, 31, 37
    ]

    for p in small_primes:
        if n == p:
            return True
        if n % p == 0:
            return False
    d = n - 1
    s = 0
    while d % 2 == 0:
        s += 1
        d //= 2

    # n长度小于64bit时使用小素数测试
    if n.bit_length() <= 64:
        test_bases = [2, 3, 5, 7, 11, 13, 17]
    else:
        test_bases = [random.randrange(2, n - 2) for _ in range(rounds)]

    for a in test_bases:

        if a % n == 0:
            continue

        x = pow(a, d, n)

        if x == 1 or x == n - 1:
            continue

        skip = False
        for _ in range(s - 1):
            x = pow(x, 2, n)
            if x == n - 1:
                skip = True
                break

        if skip:
            continue

        return False

    return True


# 素数筛,返回指定整数范围内的所有素数
def prime_sieve(limit=100000):
    sieve = bytearray(b"\x01") * (limit + 1)
    sieve[0:2] = b"\x00\x00"

    for i in range(2, int(limit ** 0.5) + 1):
        if sieve[i]:
            step = i
            start = i * i
            sieve[start:limit + 1:step] = b"\x00" * (
                ((limit - start) // step) + 1
            )

    return [i for i in range(limit + 1) if sieve[i]]


# 10w内所有素数
SMALL_PRIMES = prime_sieve(100000)


# 试除法分解较小因子
def trial_division(n: int, factors: list) -> int:

    for p in SMALL_PRIMES:
        if p * p > n:
            break

        while n % p == 0:
            factors.append(p)
            n //= p

    return n


# Pollard p-1
def pollard_p1(n: int, B1=10000) -> int:

    if n % 2 == 0:
        return 2

    a = 2

    for p in SMALL_PRIMES:
        if p > B1:
            break

        pk = p
        while pk * p <= B1:
            pk *= p

        a = pow(a, pk, n)

    d = gcd(a - 1, n)

    if 1 < d < n:
        return d

    return 1


# Pollard rho
# 这里采用的F是多项式函数F(x) = x^2 +c
def pollard_rho(n: int) -> int:

    if n % 2 == 0:
        return 2

    if n % 3 == 0:
        return 3

    while True:

        c = random.randrange(1, n - 1)
        x = random.randrange(2, n - 1)
        y = x
        d = 1

        while d == 1:
            x = (x*x + c) % n
            y = (y*y + c) % n
            y = (y*y + c) % n

            d = gcd(abs(x - y), n)

        if d != n:
            return d




def pollard_rho(n: int) -> int:
    if n % 2 == 0: return 2
    if is_prime(n): return n

    step_size = 127 # 批量 GCD 的步数
    
    while True:
        c = random.randrange(1, n)
        x = random.randrange(2, n)
        y = x
        r = 1
        q = 1
        d = 1
        
        while d == 1:
            x = y
            # r为Brent算法的增长序列指数
            for _ in range(r):
                y = (y * y + c) % n
            
            k = 0
            while k < r and d == 1:
                ys = y
                # 批量乘积阶段
                for _ in range(min(step_size, r - k)):
                    y = (y * y + c) % n
                    q = (q * abs(x - y)) % n
                
                d = gcd(q, n)
                k += step_size
            r *= 2
        
        # 如果 GCD 结果是 n，说明需要回退找具体一因子
        if d == n:
            while True:
                ys = (ys * ys + c) % n
                d = gcd(abs(x - ys), n)
                if d > 1:
                    break
        
        if d < n:
            return d

# 递归调用因子分解
def factor_recursive(n: int, factors: list):

    if n == 1:
        return

    if is_prime(n):
        factors.append(n)
        return

    # Pollard p-1
    d = pollard_p1(n)

    # 失败来到Pollard rho
    if d == 1:
        d = pollard_rho(n)

    factor_recursive(d, factors)
    factor_recursive(n // d, factors)


# 完整分解
def factor(n: int):

    start = time.perf_counter()
    factors = []

    # 首先使用试除法
    n = trial_division(n, factors)

    if n > 1:
        factor_recursive(n, factors)

    factors.sort()

    elapsed = time.perf_counter() - start
    return factors, elapsed


# 输出结果
def print_factorization(n: int):

    factors, t = factor(n)

    cnt = Counter(factors)

    parts = []
    for p in sorted(cnt):
        e = cnt[p]
        if e == 1:
            parts.append(f"{p:0x}")
        else:
            parts.append(f"{p}^{e}")
    
    # 转为16进制
    for i in range(len(factors)):
        factors[i] = f"{factors[i]:0x}"

    print("=" * 60)
    print("n =", n)
    print("Prime Factorization:")
    print(" * ".join(parts))
    print("All factors list:")
    print(factors)
    print(f"Time used: {t:.6f} seconds")
    print("=" * 60)



if __name__ == "__main__":

    tests = [
        int("22F4AE6C291EC7B8A534F0D9B13C020D26",16)
    ]

    for x in tests:
        print_factorization(x)

File: utils/Factorization/Factorizator/factorizator.cpp

#include <NTL/ZZ.h>
#include <iostream>
#include <vector>
#include <map>
#include <algorithm>
#include <chrono>
#include <string>

using namespace std;
using namespace NTL;

// =====================================================
// gcd 自实现
// =====================================================
ZZ gcdZZ(ZZ a, ZZ b) {
    while (b != 0) {
        ZZ t = a % b;
        a = b;
        b = t;
    }
    return a;
}

// =====================================================
// 十六进制字符串转 ZZ
// 支持: 0xABCDEF
// =====================================================
ZZ hex_to_ZZ(string s) {

    if (s.size() >= 2 &&
        s[0] == '0' &&
       (s[1] == 'x' || s[1] == 'X'))
        s = s.substr(2);

    ZZ x(0);

    for (char c : s) {

        x *= 16;

        if ('0' <= c && c <= '9')
            x += (c - '0');
        else if ('a' <= c && c <= 'f')
            x += (c - 'a' + 10);
        else if ('A' <= c && c <= 'F')
            x += (c - 'A' + 10);
    }

    return x;
}

// =====================================================
// Miller-Rabin 自实现
// 使用 NTL 的 PowerMod 做模幂
// =====================================================
bool miller_rabin(const ZZ& n, int rounds = 10) {

    if (n < 2) return false;
    if (n == 2 || n == 3) return true;
    if (n % 2 == 0) return false;

    // n-1 = d * 2^s
    ZZ d = n - 1;
    long s = 0;

    while (d % 2 == 0) {
        d /= 2;
        s++;
    }

    long bases[] = {2,3,5,7,11,13,17,19,23,29};

    for (int i = 0; i < rounds; i++) {

        ZZ a = conv<ZZ>(bases[i % 10]);

        if (a >= n - 2)
            a = ZZ(2);

        ZZ x = PowerMod(a, d, n);

        if (x == 1 || x == n - 1)
            continue;

        bool witness = true;

        for (long r = 1; r < s; r++) {

            x = MulMod(x, x, n);

            if (x == n - 1) {
                witness = false;
                break;
            }
        }

        if (witness)
            return false;
    }

    return true;
}

// =====================================================
// 小素数筛
// =====================================================
vector<long> get_primes(long limit = 100000) {

    vector<bool> vis(limit + 1, false);
    vector<long> primes;

    for (long i = 2; i <= limit; i++) {

        if (!vis[i]) {

            primes.push_back(i);

            if (1LL * i * i <= limit) {
                for (long j = i * i; j <= limit; j += i)
                    vis[j] = true;
            }
        }
    }

    return primes;
}

vector<long> SMALL_PRIMES = get_primes();

// =====================================================
// 试除法
// =====================================================
void trial_division(ZZ& n, vector<ZZ>& fac) {

    for (long p : SMALL_PRIMES) {

        ZZ P(p);

        if (P * P > n)
            break;

        while (n % P == 0) {
            fac.push_back(P);
            n /= P;
        }
    }
}

// =====================================================
// Pollard p-1
// =====================================================
ZZ pollard_p1(const ZZ& n, long B1 = 10000) {

    if (n % 2 == 0)
        return ZZ(2);

    ZZ a(2);

    for (long p : SMALL_PRIMES) {

        if (p > B1)
            break;

        long pk = p;

        while (1LL * pk * p <= B1)
            pk *= p;

        a = PowerMod(a, conv<ZZ>(pk), n);
    }

    ZZ d = gcdZZ(a - 1, n);

    if (d > 1 && d < n)
        return d;

    return ZZ(1);
}

// =====================================================
// Pollard Rho
// =====================================================
ZZ pollard_rho(const ZZ& n) {

    if (n % 2 == 0)
        return ZZ(2);

    while (true) {

        ZZ c = RandomBnd(n - 1) + 1;
        ZZ x = RandomBnd(n - 2) + 2;
        ZZ y = x;
        ZZ d(1);

        while (d == 1) {

            x = (MulMod(x, x, n) + c) % n;

            y = (MulMod(y, y, n) + c) % n;
            y = (MulMod(y, y, n) + c) % n;

            d = gcdZZ(abs(x - y), n);
        }

        if (d != n)
            return d;
    }
}

// =====================================================
// 递归分解
// =====================================================
void factor_rec(const ZZ& n, vector<ZZ>& fac) {

    if (n == 1)
        return;

    if (miller_rabin(n)) {
        fac.push_back(n);
        return;
    }

    ZZ d = pollard_p1(n);

    if (d == 1)
        d = pollard_rho(n);

    factor_rec(d, fac);
    factor_rec(n / d, fac);
}

// =====================================================
// 总分解器
// =====================================================
vector<ZZ> factor_all(ZZ n) {

    vector<ZZ> fac;

    trial_division(n, fac);

    if (n > 1)
        factor_rec(n, fac);

    sort(fac.begin(), fac.end());

    return fac;
}

// =====================================================
// 输出
// =====================================================
void print_result(const ZZ& n) {

    auto start = chrono::high_resolution_clock::now();

    vector<ZZ> fac = factor_all(n);

    auto end = chrono::high_resolution_clock::now();

    map<ZZ,int> cnt;

    for (auto &x : fac)
        cnt[x]++;

    cout << "====================================\n";
    cout << "n = " << n << "\n";
    cout << "Prime factorization:\n";

    bool first = true;

    for (auto &kv : cnt) {

        if (!first) cout << " * ";
        first = false;

        cout << kv.first;

        if (kv.second > 1)
            cout << "^" << kv.second;
    }

    cout << "\n";

    double t =
        chrono::duration<double>(end - start).count();

    cout << "Time used: " << t << " sec\n";
    cout << "====================================\n";
}

// =====================================================
// main
// =====================================================
int main() {

    string s = "0x22F4AE6C291EC7B8A534F0D9B13C020D26";

    cout<<"The BIG num is : "<<s<<endl;

    ZZ n;

    if (s.size() >= 2 &&
        s[0] == '0' &&
       (s[1] == 'x' || s[1] == 'X'))
        n = hex_to_ZZ(s);
    else
        n = conv<ZZ>(s.c_str());

    print_result(n);

    return 0;
}

